document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -386,28 +387,58 @@ public long total(Query query) {
   @Override
   public boolean bulkUpsert(Map<Key, Document> documents) {
     try {
-      List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
-      for (Entry<Key, Document> entry : documents.entrySet()) {
-        Key key = entry.getKey();
-        // insert or overwrite
-        bulkCollection.add(new UpdateOneModel<>(
-            this.selectionCriteriaForKey(key),
-            prepareUpsert(key, entry.getValue()),
-            new UpdateOptions().upsert(true)));
-      }
-
-      BulkWriteResult result = Failsafe.with(bulkWriteRetryPolicy)
-          .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+      BulkWriteResult result = bulkUpsertImpl(documents);
       LOGGER.debug(result.toString());
-
       return true;
-
     } catch (IOException | MongoServerException e) {
       LOGGER.error("Error during bulk upsert for documents:{}", documents, e);
       return false;
     }
   }
 
+  private BulkWriteResult bulkUpsertImpl(Map<Key, Document> documents) throws JsonProcessingException {
+    List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
+    for (Entry<Key, Document> entry : documents.entrySet()) {
+      Key key = entry.getKey();
+      // insert or overwrite
+      bulkCollection.add(new UpdateOneModel<>(
+          this.selectionCriteriaForKey(key),
+          prepareUpsert(key, entry.getValue()),
+          new UpdateOptions().upsert(true)));
+    }
+
+    return Failsafe.with(bulkWriteRetryPolicy)
+        .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+  }
+
+  @Override
+  public Iterator<Document> bulkUpsertAndReturnOlderDocuments(Map<Key, Document> documents) throws IOException {
+    try {
+      // First get all the documents for the given keys.
+      FindIterable<BasicDBObject> cursor = collection.find(selectionCriteriaForKeys(documents.keySet()));
+      final MongoCursor<BasicDBObject> mongoCursor = cursor.cursor();
+
+      // Now go ahead and do the bulk upsert.
+      BulkWriteResult result = bulkUpsertImpl(documents);
+      LOGGER.debug(result.toString());
```

Comment:
I am not sure about the performance of this. Not a Javaer here but it seams whether logger is enabled or not we still turn it into string? cc @kotharironak

Reply:
👍 we should always be letting the logger doing the stringification for us so we don't have to eat this cost unless the message is needed. That means wrapping it in an `if` or IMO, more graceful to do `LOGGER.debug("{}", result);`

Reply:
Fixing in a new PR. Thanks!

Reply:
https://github.com/hypertrace/document-store/pull/25